### ISSUE TITLE
In the release environment, monitoring data changes does not work

### ISSUE THREAD

**Issue by Tang-Hai** [realm/realm-dart#1286]
### What happened?

It is normal to trigger a callback after monitoring data changes in the test environment, but after packaging into a release environment, the application monitoring data changes does not work

### Repro steps

final all = DataDB.mainRealm.all<DataUser>();
    final user = all.firstWhere((a) => a.isDefault);
    this.user = user.obs;
    chatSubscription = user.useLimit?.chat.changes.listen((event) async {
      // After the data changes, the test environment executes, and the release environment does not execute
    });

### Version

 realm: ^1.0.3, Flutter is already up to date on channel stable Flutter 3.10.1 • channel stable • https://github.com/flutter/flutter.git Framework • revision d3d8effc68 (5 days ago) • 2023-05-16 17:59:05 -0700 Engine • revision b4fb11214d Tools • Dart 3.0.1 • DevTools 2.23.1

### What Atlas Services are you using?

Local Database only

### What type of application is this?

Flutter Application

### Client OS and version

android all OS and version

### Code snippets

final all = DataDB.mainRealm.all<DataUser>();
    final user = all.firstWhere((a) => a.isDefault);
    this.user = user.obs;
    chatSubscription = user.useLimit?.chat.changes.listen((event) async {
      // After the data changes, the test environment executes, and the release environment does not execute
    });

### Stacktrace of the exception/crash you're getting

```shell
No exceptions, only problems with release
```

### Relevant log output

```shell
no log
```

**Comment by dotjon0** [comment]
No directly related, but Flutter 3.10.0 seems to have broken many packages on pub dev... For example Firebase Auth has stopped working and if you search in Google it seems lots of packages have been broken... If you need a fast fix, downgrading Flutter has worked for other packages as a temp measure...so may be worth a spin in the meantime...

**Comment by Tang-Hai** [comment]
> No directly related, but Flutter 3.10.0 seems to have broken many packages on pub dev... For example Firebase Auth has stopped working and if you search in Google it seems lots of packages have been broken... If you need a fast fix, downgrading Flutter has worked for other packages as a temp measure...so may be worth a spin in the meantime...

Thank you for your reply, so which version should I downgrade to?

**Comment by nielsenko** [comment]
@Tang-Hai This slipped between chairs. Do you still have an issue here?

**Comment by tanguy-penfen** [comment]
Hello,
I have the same issue : suscribing to changes on a RealmResults works fine in debug but doesn't show any events after the first one (at init) in release mode.

**Comment by dotjon0** [comment]
> > No directly related, but Flutter 3.10.0 seems to have broken many packages on pub dev... For example Firebase Auth has stopped working and if you search in Google it seems lots of packages have been broken... If you need a fast fix, downgrading Flutter has worked for other packages as a temp measure...so may be worth a spin in the meantime...
> 
> Thank you for your reply, so which version should I downgrade to?

sorry did not see this. If its macOS or iOS and you have not configured Keychain Sharing capability enabled, you can see a temporary fix here https://github.com/firebase/flutterfire/issues/11186#issuecomment-1605589309 until Keychain Sharing capability is enabled - use flutter 3.3.9. Although Firebase Auth released this back in Q1, this only hit us when we upgraded to 3.10.0 for some reason, but this is the root cause.

**Comment by nielsenko** [comment]
@tanguy-penfen 
I'm a bit surprised that there should be any difference in behaviour between:
```dart
class MyClass {
  late final realmResults = realm.all<Item>();
  late final streamSubscription;
  MyClass(){
    streamSubscription = reamResults.changes.listen((event){...});
}
```
and
```dart
class MyClass {
  late final realmResults;
  late final streamSubscription;
  MyClass(){
    realmResults = realm.all<Item>();
    streamSubscription = reamResults.changes.listen((event){...});
}
```
This does not match my understanding of when `late final` initialisers are called. It should behave exactly the same.

**Comment by nielsenko** [comment]
Still does not compute for me. The `realmResults` member will not be stripped by tree-shaking. Are you saying you can re-create this issue by switching to the ` late final realmResults = realm.all<Item>();` version?

**Comment by leandrorazevedo** [comment]
same issue here

**Comment by leandrorazevedo** [comment]
Hello everyone, I'm still facing the same issue.
I've already updated to the latest version, and the behavior is still the same.
In debug mode, the synchronization is working perfectly. However, when I generate an APK in release mode and test it on the device, the synchronization doesn't work; it doesn't seem to receive any data.
After I close the app and open it again, the information gets updated.

Please, I wouldn't like to move to another solution/platform.

**Comment by nielsenko** [comment]
@leandrorazevedo If you can provide us with a minimal reproduction, or share your code, then we can take a look. I promise you that notifications work in release mode as well.

Otherwise, I would look for stream subscriptions that are not tracked and hence streams that become eligible for GC before you intend it. In Dart the GC reachability analysis is a fair bit more aggressive in release than in debug mode.

**Comment by leandrorazevedo** [comment]
@nielsenko, I'm sorry to insist on posting here in this thread, but I believe it can be useful for other programmers...

After conducting several tests, I've come to a conclusion (which doesn't make sense to me, but I'll leave it here)...

Just a reminder of the situation: The realm's 'listen' works in debug mode and doesn't work in release mode.

```dart
class _HomeScreenState extends State<HomeScreen> {

    RxInt qtNotificationNotRead = 0.obs;
    late final StreamSubscription<RealmResultsChanges<schema.Notification>> streamSubscription;

    @override
    void initState() {
       super.initState();
       _initNotificationRealm();
    }

    Future<void> _initNotificationRealm() async {
        streamSubscription = allNotificationsRealm!.all<schema.Notification>().changes.listen((event) {
          qtNotificationNotRead.value = event.results.where((element) => !element.isRead).length;
        });
    }

    @override
    Widget build(BuildContext context) {
       ...
       icon: Obx(
                      () => Badge(
                          label: BodyText.small(qtNotificationNotRead.value > 9 ? "9+" : ${qtNotificationNotRead.value}", color: colorWhite),
                          isLabelVisible: qtNotificationNotRead.value > 0,
                          backgroundColor: colorError,
                          child: const SvgWidget('regular/bell', color: colorWhite),
                      ),
                 ),
       ...
    }
}
```

The code above works perfectly fine in both the debug and release environments.

But if we change ONLY the variable declaration to (excluding final):

```dart
    late StreamSubscription<RealmResultsChanges<schema.Notification>> streamSubscription;
```
The app's behavior starts to match the description in the title of this thread: the 'listen' method gets triggered in the debug environment but doesn't get triggered in a release mode APK.

I hope that this can help other people in some way.

**Comment by nielsenko** [comment]
Why is `_initNotification` **async**?

**Comment by leandrorazevedo** [comment]
It was just an error in the text formatting for the issue.
Here's the complete code: (but I think this doesn't interfere with our issue)

```dart
  Future<void> _initNotificationRealm() async {
    if (userController.user != null) {
      allNotificationsRealm = await RealmUtils.createNotificationDb(userId: userController.user!.id);

      streamSubscription = allNotificationsRealm!.all<schema.Notification>().changes.listen((event) {
          qtNotificationNotRead.value = event.results.where((element) => !element.isRead).length;
        });
    }
  }
```

**Comment by leandrorazevedo** [comment]
Here's my implementation for the `createNotificationDb` method:

```dart
enum CollectionType { ctNotifications }

class RealmUtils {
  static Future<Realm> createNotificationDb({required String userId}) async {
    const collectionType = CollectionType.ctNotifications;

    final user = await _createAppUser();

    final flxConfig = Configuration.flexibleSync(user, [Notification.schema], path: await _absolutePath("db_${collectionType.name}.realm"));

    var realm = Realm(flxConfig);
    debugPrint("Created local realm db at: ${realm.config.path}");

    final RealmResults<Notification> query = realm.query<Notification>('owner == $userId && deleted == false');

    realm.subscriptions.update((mutableSubscriptions) {
      mutableSubscriptions.add(query);
    });

    realm.subscriptions.waitForSynchronization().then((value) {
      debugPrint("Synchronization completed for realm: ${realm.config.path}");
    });

    return realm;
  }

  static Future<User> _createAppUser() async {
    const String appId = "xxxxxxxxxxxxxx";
    final appConfig = AppConfiguration(appId);
    final app = App(appConfig);
    final apiKeyCredentials = Credentials.apiKey("xxxxxxxxxxxxxxxx");
    final user = await app.logIn(apiKeyCredentials);
    return user;
  }

  static Future<String> _absolutePath(String fileName) async {
    final appDocsDirectory = await getApplicationDocumentsDirectory();
    final realmDirectory = '${appDocsDirectory.path}/mongodb-realm';
    if (!Directory(realmDirectory).existsSync()) {
      await Directory(realmDirectory).create(recursive: true);
    }
    return "$realmDirectory/$fileName";
  }
}

```

**Comment by leandrorazevedo** [comment]
> Looking at createNotificationDb I do wonder why you are opening a realm that is named after the collection type? Or maybe you really only have that one realm type in your app?

At the moment, we only have this realm type.
We are still conducting analysis and testing of the solution.
We have already made significant progress. In our architecture, the use of flexible sync (app sync) is crucial.

Regarding platforms, we are working with Android and iOS.

**Comment by MainasuK** [comment]
My little project has the same issue in the release build on Windows & macOS. 

Thanks for @leandrorazevedo providing the `_initNotificationRealm` method to work around this issue. 

The project is open-source and the problem was fixed in v1.2.1.
> v.1.2.1 download [here](https://github.com/MainasuK/train_map/actions/runs/7887994544/artifacts/1241720141)
> v1.2.0 download [here](https://github.com/MainasuK/train_map/actions/runs/7865405713/artifacts/1237005778)
> Project source code: https://github.com/MainasuK/train_map

Maybe that's enough info to help you take a deep look at what's wrong. @nielsenko 

```
flutter --version
Flutter 3.16.7 • channel stable • https://github.com/flutter/flutter.git
Framework • revision ef1af02aea (5 weeks ago) • 2024-01-11 15:19:26 -0600
Engine • revision 4a585b7929
Tools • Dart 3.2.4 • DevTools 2.28.5
```

How to reproduce the issue:
1. Open the app and welcome page display
2. Copy sheet A content and paste it into the first text input (use text editor e.g VS Code)
3. Copy sheet B content and paste it into the second text input
4. Scroll to the bottom of the page. Click the button to import the data

Expect: The sidebar should update and a new data entry row appear
Actual: Nothing updated. Close and open the app again. The data entry display in the sidebar

[data_sheet.zip](https://github.com/realm/realm-dart/files/14267951/data_sheet.zip)

**Comment by nirinchev** [comment]
@MainasuK this issue has been closed for 6 months - if you have a similar problem, please file a new issue and fill in the template.
